src/Toolbox/Util/Graph.hs

```diff
@@ -0,0 +1,207 @@
+          os' = L.nub $ L.sort $ (os ++ concatMap fst newdeps)
+          is' = L.nub $ L.sort $ (is ++ concatMap snd newdeps)
+       in (c, (os', is'))
+    clusteredGraph' = fmap merge clusteredGraph
+    unclustered = clusterStateUnclustered clustering
+    unclusteredGraph' = filter (\(v, _) -> v `elem` unclustered) unclusteredGraph
+
+replaceStep ::
+  ClusterState ->
+  GraphState ->
+  GraphState
+replaceStep clustering (GraphState clusteredGraph unclusteredGraph) = GraphState clusteredGraph' unclusteredGraph'
+  where
+    replaceV c@(Clustered {}) = c
+    replaceV (Unclustered v) =
+      case matchCluster clustering v of
+        Nothing -> Unclustered v
+        Just cls -> Clustered cls
+    --
+    replace :: (a -> ICVertex -> Bool) -> (a, ([ICVertex], [ICVertex])) -> (a, ([ICVertex], [ICVertex]))
+    replace eq (v, (os, is)) =
+      let os' = filter (not . eq v) $ L.nub $ L.sort $ fmap replaceV os
+          is' = filter (not . eq v) $ L.nub $ L.sort $ fmap replaceV is
+       in (v, (os', is'))
+    --
+    clusteredGraph' = fmap (replace ((==) . Clustered)) clusteredGraph
+    unclusteredGraph' = fmap (replace (\_ _ -> False)) unclusteredGraph
+
+clusterStep ::
+  GraphState ->
+  ClusterState ->
+  ClusterState
+clusterStep (GraphState clusteredGraph _) clustering = flip execState clustering $ do
+  clustered' <- traverse go clustered
+  modify' (\s -> s {clusterStateClustered = clustered'})
+  where
+    clustered = clusterStateClustered clustering
+    go (cls, vs) = do
+      unclustered <- clusterStateUnclustered <$> get
+      let (_os, is) = fromMaybe ([], []) $ L.lookup cls clusteredGraph
+          -- NOTE: only cluster towards upper dependency.
+          is' = getUnclustered is
+          added = (is' L.\\ vs) `L.intersect` unclustered
+          vs' = L.nub $ L.sort (vs ++ added)
+          unclustered' = L.nub $ L.sort (unclustered L.\\ added)
+      modify' (\s -> s {clusterStateUnclustered = unclustered'})
+      pure (cls, vs')
+
+fullStep ::
+  (ClusterState, GraphState) ->
+  (ClusterState, GraphState)
+fullStep (clustering, graphState) =
+  let graphState1 = mergeStep clustering graphState
+      graphState2 = replaceStep clustering graphState1
+      clustering' = clusterStep graphState2 clustering
+   in clustering' `seq` graphState2 `seq` (clustering', graphState2)
+
+mkRevDep :: [(Int, [Int])] -> IntMap [Int]
+mkRevDep deps = L.foldl' step emptyMap deps
+  where
+    emptyMap = L.foldl' (\(!acc) (i, _) -> IM.insert i [] acc) IM.empty deps
+    step !acc (i, js) =
+      L.foldl' (\(!acc') j -> IM.insertWith (<>) j [i] acc') acc js
+
+nodeSizeLimit :: Int
+nodeSizeLimit = 150
+
+-- ( vertex, (dep, revdep)) per each item
+getBiDepGraph :: ModuleGraphInfo -> [(Int, ([Int], [Int]))]
+getBiDepGraph graphInfo =
+  let modDep = mginfoModuleDep graphInfo
+      modRevDepMap = mkRevDep modDep
+      modRevDep = IM.toList modRevDepMap
+      modBiDep = concat $ inner grouping joiner fst fst modDep modRevDep
```

Comment:
I'm a little bit baffled about why we need the `discrimination` library here. Are we not simply doing a `Map` aggregation? Is `discrimination` significantly faster than `Data.Map.Map`?

Reply:
Hmmm, the docs for `inner` say it's `O(n)`, whereas doing a map aggregation would be `Omega(n * log n)` and more likely `O(n * log n * log n)` since you have to insert things into `Set`s.

Reply:
`Data.Map.Map` would also use `O(n * n)` memory just to store the map itself, huh? I wonder what `discrimination`'s memory footprint is like.

Reply:
yes. i guess `discrimination` should have performance benefits as it uses radix sort and aimed at linear time joining operation. 
in fact, I am not very careful about asymptotic performance characteristic here (in the end, the use of list instead of set is totally suboptimal), but i used the library as it provides generic interface for joining.
I will add some comment about that as this is of general interest. Thanks for the discussion on that aspect!